### PR TITLE
PCRM-3568-3569-Test-Coverage

### PR DESCRIPTION
**Pull request by premnhs**

### CODE REVIEW

File: test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                        if ("tbc-OnlyOnGP" in str(filename)) :
+                            
+                            data = s3.get_object(Bucket=LR_13_BUCKET, Key=key_items.get("Key"))
+                            act_contents = sorted(data["Body"].read().decode("utf-8").splitlines())
+                            for line, row in zip(act_contents,exp_gponly_data):
+                                
+                                if row.rstrip() == line:
+                                        Messages.write_message("Actual row is :" + str(line))
+                                        Messages.write_message("success : Record as expected")
+                                        assert True
```
Comment (jamesnunnnhs) [inline test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py:81]: I think I've mentioned before but using `assert True` has no use whatsoever, it doesn't do anything behind the scenes. You could just leave it out.

File: test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    
+    s3 = dev.client("s3", region_name)
+    result = s3.list_objects(
+        Bucket=LR_13_BUCKET
+        )
+    with open(exp_gponly_path, 'r') as exp_gponly_datafile:
+        exp_gponly_data = sorted(exp_gponly_datafile)
+        
+        if result.get("Contents") != None:
+                for key_items in result.get("Contents"):
```
Comment (jamesnunnnhs) [inline test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py:70]: This indentation seems a bit odd. All indentation should be 4 spaces

File: test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                            for line, row in zip(act_contents,exp_gponly_data):
+                                
+                                if row.rstrip() == line:
+                                        Messages.write_message("Actual row is :" + str(line))
+                                        Messages.write_message("success : Record as expected")
+                                        assert True
+                                else :
+                                        Messages.write_message("Actual row is :" + str(line) + "  Expected was :" + row.rstrip())
+                                        Messages.write_message("Unsuccessful : Record as expected")
+                                        assert False
```
Comment (jamesnunnnhs) [inline test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py:85]: Can you not just do:
```
assert row.rstrip() == line, "Actual row is :" + str(line) + " Expected was :" + row.rstrip()\nUnsuccessful : Record as expected"

```

File: lambdas/LR_12_pds_registration_status/pds_registration_status.py [hunk 4]
```diff
@@ -82,6 +82,7 @@ def get_pds_exclusive_registrations(job_id: str) -> Dict[str, str]:
         if nhs_number not in job_nhs_numbers:
             try:
                 pds_record = get_pds_record(nhs_number, max_retries=PDS_API_RETRIES)
+                pds_record["address"] = (pds_record["address"] + [None, None, None, None, None])[:5]
```
[the remark this answers is not in the document]
Comment (premnhs) [reply lambdas/LR_12_pds_registration_status/pds_registration_status.py:85]: This was a fix from James which is required for failing tests.

File: test/integrationtests/step_impl/lr_09_jobscheduler_check.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+
+@step("trigger lr09 and expected statuscode is <expstatuscode>")
+def trigger_lr09(expstatuscode):
+    d = {}
+    client = dev.client('lambda', region_name="eu-west-2")
+    response = client.invoke(
+    FunctionName=LR_09_LAMBDA_ARN,
+    LogType='Tail',
+    Payload= json.dumps(d)
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_09_jobscheduler_check.py:29]: missing indent

File: test/integrationtests/step_impl/lr_09_jobscheduler_check.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+region_name = "eu-west-2"
+test_datetime = get_datetime_now()
+
+aws_resource = get_aws_resources()
+LR_09_LAMBDA_ARN = aws_resource["lr_09_lambda"]['value'] 
+
+
+@step("trigger lr09 and expected statuscode is <expstatuscode>")
+def trigger_lr09(expstatuscode):
+    d = {}
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_09_jobscheduler_check.py:24]: This is only used as a json.dumps, is it actually assigned to? seems a bit irrelevant

File: test/integrationtests/step_impl/lr_09_jobscheduler_check.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    client = dev.client('lambda', region_name="eu-west-2")
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_09_jobscheduler_check.py:25]: `region_name=region_name` defined above

File: test/integrationtests/step_impl/lr_09_jobscheduler_check.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+from utils.datetimezone import get_datetime_now
+from .tf_aws_resources import get_aws_resources
+
+# On github
+access_key = os.getenv("AWS_PUBLIC_KEY")
+secret_key = os.getenv("AWS_PRIVATE_KEY")
+dev = boto3.session.Session(access_key, secret_key)
+
+region_name = "eu-west-2"
+test_datetime = get_datetime_now()
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_09_jobscheduler_check.py:16]: Globals should be in caps

File: test/integrationtests/step_impl/lr_09_jobscheduler_check.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+@step("trigger lr09 and ensure scheduled checked successfully completed")
+def trigger_lr09_get_requestid():
+    d = {}
+    client = dev.client('lambda', region_name="eu-west-2")
+    response = client.invoke(
+    FunctionName=LR_09_LAMBDA_ARN,
+    LogType='Tail',
+    Payload= json.dumps(d)
+    )
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_09_jobscheduler_check.py:44]: Same comments for this as above

File: test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    integ_auto_test = ("integ_auto_test_"+ time_gen)
+    lr_10_response = client.start_execution(
+        stateMachineArn=LR_10_STATE_FUNCTION_ARN,
+        name=integ_auto_test,
+        input=json.dumps({"job_id": job_id})
+    )
+    time.sleep(5)
+    executionarn = lr_10_response["executionArn"]
+    desc_exec_resp = client.describe_execution(executionArn=executionarn)
+    status = desc_exec_resp['status']
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py:42]: Could you split this out a bit? it's currently a big block of scary code :D

File: .github/workflows/create-feature-branch-infra.yml [hunk 11]
```diff
@@ -20,7 +20,7 @@ jobs:
           make python-package
       - name: Terraform Format Check
         run: |
-          make fmt-check
+          make fmt
```
[the remark this answers is not in the document]
Comment (joeds13) [reply .github/workflows/create-feature-branch-infra.yml:23]: No, this is telling us that the provide `terraform` files in this commit are not formatted as expected

This needs reverting, then `make fmt` running locally and the files that get changed adding and committing to this PR

File: test/integrationtests/step_impl/lr_beforehooks.py [hunk 12]
```diff
@@ -4,21 +4,42 @@
...
 s3 = dev.client("s3")
 LR_01_BUCKET = aws_resource['lr_01_bucket']['value']
+LR_13_BUCKET = aws_resource["lr_13_bucket"]["value"]
+INFLIGHT_TABLE = aws_resource['inflight_table']['value']
 
 
-@step("setup steps: clear all files in lr01 folders")
+@step("setup steps: clear all files in bucket folders")
 def before_hooks():
+ # BUCKETS
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_beforehooks.py:20]: Wrong indenting. Also does it need to be in all caps? :D

File: test/integrationtests/step_impl/lr_beforehooks.py [hunk 13]
```diff
@@ -4,21 +4,42 @@
...
+            s3.delete_object(Bucket=LR_01_BUCKET, Key=object['Key'])
+    
+    lr_13_response = s3.list_objects_v2(Bucket=LR_13_BUCKET)
+
+    if 'Contents' in lr_13_response :
+        for object in lr_13_response['Contents']:
+            print('Deleting', object['Key'])
+            s3.delete_object(Bucket=LR_13_BUCKET, Key=object['Key'])
+
+#DynamoDB
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_beforehooks.py:35]: Wrong indenting

File: .github/workflows/validate-terraform-list-rec-stack.yml [hunk 14]
```diff
@@ -20,7 +20,7 @@ jobs:
           make python-package
       - name: Terraform Format Check
         run: |
-          make fmt-check
+          make fmt
```
Comment (joeds13) [inline .github/workflows/validate-terraform-list-rec-stack.yml:23]: This needs reverting, then make fmt running locally and the files that get changed adding and committing to this PR

File: test/integrationtests/step_impl/lr_09_jobscheduler_check.py (hunk 6 above)
Comment (premnhs) [reply test/integrationtests/step_impl/lr_09_jobscheduler_check.py:24]: its used here in the json.dumps,
`client.invoke(
        FunctionName=LR_09_LAMBDA_ARN, LogType="Tail", Payload=json.dumps(d)
    )`

Comment (josh-scargill-nhsd) [reply test/integrationtests/step_impl/lr_09_jobscheduler_check.py:22]: But wouldn't it work with `client.invoke( FunctionName=LR_09_LAMBDA_ARN, LogType="Tail", Payload=json.dumps({}) )` ?

Comment (premnhs) [reply test/integrationtests/step_impl/lr_09_jobscheduler_check.py:22]: yeah my bad

File: test/integrationtests/step_impl/lr_09_jobscheduler_check.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        FunctionName=LR_09_LAMBDA_ARN, LogType="Tail", Payload=json.dumps({})
+    )
+    for key, value in response.items():
+        if key == "ResponseMetadata":
+            assert response["ResponseMetadata"]["HTTPStatusCode"] == int(expstatuscode)
+
+
+@step("trigger lr09 and ensure scheduled checked successfully completed")
+def trigger_lr09_get_requestid():
+    d = {}
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_09_jobscheduler_check.py:33]: 👀

File: test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+                                Messages.write_message("Actual row is :" + str(line))
+                                Messages.write_message("success : Record as expected")
+                            else:
+                                assert row.rstrip() == line, (
+                                    "Actual row is :"
+                                    + str(line)
+                                    + " Expected was :"
+                                    + row.rstrip()
+                                    + "\nUnsuccessful : Record as expected"
+                                )
```
Comment (josh-scargill-nhsd) [inline test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py:104]: Could you not use string interpolation instead? Thinking it could be easier to read like
`assert row.rstrip() == line, f"Actual row is : {str(line)} Expected was : {row.rstrip()}\nUnsuccessful : Record as expected"`

Comment (premnhs) [reply test/integrationtests/step_impl/lr_10_11_13_onlyon_gppractice.py:104]: it was the same before..not sure if black intend updated this ... reverted it back now

**Review by josh-scargill-nhsd — APPROVED** [review]

**Review by HarveyNHS — APPROVED** [review]
Approved

**Review by jamesnunnnhs — APPROVED** [review]
